File: Assets/Scripts/Core/Game/ObjectKinds.cs
using System;
using System.Collections.Generic;

namespace CyberRider.Core
{
    public sealed class Placement
    {
        public EntityDef Entity;
        public PropDef Prop;
        public List<LevelLine> Lines;
    }

    public sealed class ObjectKindDef
    {
        public string Id;
        public string Name;
        public string Icon;
        /// <summary>"toys", "hazards" or "props".</summary>
        public string Group;
        public string Description;
        public Func<double, double, Placement> Make;
    }

    /// <summary>Everything the player can drop into a free-ride track, with sensible default sizes.</summary>
    public static class ObjectKinds
    {
        public static readonly ObjectKindDef[] All =
        {
            new ObjectKindDef { Id = "spring", Name = "Spring pad", Icon = "^", Group = "toys", Description = "A bouncy surface.", Make = (x, y) => new Placement { Lines = new List<LevelLine> { new LevelLine(x - 25, y, x + 25, y, "spring") } } },
            new ObjectKindDef { Id = "ramp", Name = "Ramp", Icon = "/", Group = "toys", Description = "A kicker with a vertical back.", Make = (x, y) => new Placement { Lines = new List<LevelLine> { new LevelLine(x - 40, y, x + 20, y - 30, "normal"), new LevelLine(x + 20, y - 30, x + 20, y, "normal") } } },
            new ObjectKindDef { Id = "seesaw", Name = "Seesaw", Icon = "=", Group = "toys", Description = "Tilts under weight.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "seesaw", X = x, Y = y, Length = 140 } } },
            new ObjectKindDef { Id = "pendulum", Name = "Swing", Icon = "|", Group = "toys", Description = "A bar swinging from a pivot above.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "pendulum", X = x, Y = y - 160, Length = 160, Width = 40, Amplitude = 0.8, Period = 180 } } },
            new ObjectKindDef { Id = "fanUp", Name = "Fan (up)", Icon = "^^", Group = "toys", Description = "Column of air pushing up.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "fan", X = x - 40, Y = y - 260, W = 80, H = 260, Fx = 0, Fy = -0.4 } } },
            new ObjectKindDef { Id = "fanRight", Name = "Fan (right)", Icon = ">>", Group = "toys", Description = "Air pushing to the right.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "fan", X = x, Y = y - 60, W = 220, H = 120, Fx = 0.25, Fy = -0.02 } } },
            new ObjectKindDef { Id = "magnet", Name = "Magnet", Icon = "o", Group = "toys", Description = "Pulls the rider in.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "magnet", X = x, Y = y, Radius = 160, Strength = 0.08 } } },
            new ObjectKindDef { Id = "cannon", Name = "Cannon", Icon = "*", Group = "toys", Description = "Launches at 45 degrees.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "cannon", X = x, Y = y, Angle = -45, Power = 11 } } },
            new ObjectKindDef { Id = "balloon", Name = "Balloon", Icon = "O", Group = "toys", Description = "Lifts the rider for three seconds.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "balloon", X = x, Y = y, Lift = 0.3, Duration = 120 } } },
            new ObjectKindDef { Id = "platformV", Name = "Lift", Icon = "v^", Group = "toys", Description = "Platform moving up and down.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "platform", X = x, Y = y, W = 70, Dx = 0, Dy = 60, Period = 200 } } },
            new ObjectKindDef { Id = "platformH", Name = "Shuttle", Icon = "<>", Group = "toys", Description = "Platform moving side to side.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "platform", X = x, Y = y, W = 70, Dx = 90, Dy = 0, Period = 220 } } },
            new ObjectKindDef { Id = "wall", Name = "Breakable wall", Icon = "#", Group = "hazards", Description = "Shatters when hit fast.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "wall", X1 = x, Y1 = y - 40, X2 = x, Y2 = y + 40, Threshold = 4 } } },
            new ObjectKindDef { Id = "gear", Name = "Gear", Icon = "@", Group = "hazards", Description = "Spinning toothed wheel.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "gear", X = x, Y = y, Radius = 60, Teeth = 8, Speed = 0.02 } } },
            new ObjectKindDef { Id = "train", Name = "Train", Icon = "[]", Group = "hazards", Description = "Shuttles back and forth.", Make = (x, y) => new Placement { Entity = new EntityDef { Type = "train", X = x, Y = y, W = 110, H = 28, Speed = 2, Range = 300 } } },
            new ObjectKindDef { Id = "domino", Name = "Domino", Icon = "|", Group = "props", Description = "Knock it over.", Make = (x, y) => new Placement { Prop = new PropDef { Kind = "domino", X = x, Y = y - 12, Width = 4, Height = 24, Mass = 0.5 } } },
            new ObjectKindDef { Id = "crate", Name = "Crate", Icon = "X", Group = "props", Description = "A pushable box.", Make = (x, y) => new Placement { Prop = new PropDef { Kind = "crate", X = x, Y = y - 10, Width = 20, Height = 20, Mass = 1.5 } } },
            new ObjectKindDef { Id = "tnt", Name = "TNT", Icon = "!", Group = "props", Description = "Explodes on a hard hit.", Make = (x, y) => new Placement { Prop = new PropDef { Kind = "tnt", X = x, Y = y - 9, Width = 18, Height = 18, Mass = 1, Explosive = true } } },
            new ObjectKindDef { Id = "boulder", Name = "Boulder", Icon = "()", Group = "props", Description = "Heavy and rolls.", Make = (x, y) => new Placement { Prop = new PropDef { Kind = "boulder", X = x, Y = y - 14, Radius = 14, Mass = 6 } } },
            new ObjectKindDef { Id = "cart", Name = "Cart", Icon = "__", Group = "props", Description = "Low, slippery, fast.", Make = (x, y) => new Placement { Prop = new PropDef { Kind = "cart", X = x, Y = y - 6, Width = 34, Height = 12, Mass = 1.2, Friction = 0.05 } } },
        };

        public static ObjectKindDef Get(string id)
        {
            foreach (var k in All) if (k.Id == id) return k;
            return null;
        }
    }
}

File: src/game/objectKinds.ts
import type { EntityDef } from './entities';
import type { PropDef } from './level';
import type { LineInput } from './track';

export type ObjectKind =
  | 'spring'
  | 'ramp'
  | 'seesaw'
  | 'pendulum'
  | 'fanUp'
  | 'fanRight'
  | 'magnet'
  | 'cannon'
  | 'wall'
  | 'balloon'
  | 'platformV'
  | 'platformH'
  | 'gear'
  | 'train'
  | 'domino'
  | 'crate'
  | 'tnt'
  | 'boulder'
  | 'cart';

export interface Placement {
  entity?: EntityDef;
  prop?: PropDef;
  lines?: LineInput[];
}

export interface ObjectKindDef {
  id: ObjectKind;
  name: string;
  icon: string;
  group: 'toys' | 'hazards' | 'props';
  description: string;
  make: (x: number, y: number) => Placement;
}

/** Everything the player can drop into a free-ride track, with sensible default sizes. */
export const OBJECT_KINDS: ObjectKindDef[] = [
  { id: 'spring', name: 'Spring pad', icon: '⤒', group: 'toys', description: 'A bouncy surface.', make: (x, y) => ({ lines: [{ x1: x - 25, y1: y, x2: x + 25, y2: y, material: 'spring' }] }) },
  {
    id: 'ramp',
    name: 'Ramp',
    icon: '◢',
    group: 'toys',
    description: 'A kicker with a vertical back.',
    make: (x, y) => ({
      lines: [
        { x1: x - 40, y1: y, x2: x + 20, y2: y - 30, material: 'normal' },
        { x1: x + 20, y1: y - 30, x2: x + 20, y2: y, material: 'normal' },
      ],
    }),
  },
  { id: 'seesaw', name: 'Seesaw', icon: '⚖', group: 'toys', description: 'Tilts under weight.', make: (x, y) => ({ entity: { type: 'seesaw', x, y, length: 140 } }) },
  { id: 'pendulum', name: 'Swing', icon: '⟟', group: 'toys', description: 'A bar swinging from a pivot above.', make: (x, y) => ({ entity: { type: 'pendulum', x, y: y - 160, length: 160, width: 40, amplitude: 0.8, period: 180 } }) },
  { id: 'fanUp', name: 'Fan (up)', icon: '⇈', group: 'toys', description: 'Column of air pushing up.', make: (x, y) => ({ entity: { type: 'fan', x: x - 40, y: y - 260, w: 80, h: 260, fx: 0, fy: -0.4 } }) },
  { id: 'fanRight', name: 'Fan (right)', icon: '⇉', group: 'toys', description: 'Air pushing to the right.', make: (x, y) => ({ entity: { type: 'fan', x, y: y - 60, w: 220, h: 120, fx: 0.25, fy: -0.02 } }) },
  { id: 'magnet', name: 'Magnet', icon: '◉', group: 'toys', description: 'Pulls the rider in.', make: (x, y) => ({ entity: { type: 'magnet', x, y, radius: 160, strength: 0.08 } }) },
  { id: 'cannon', name: 'Cannon', icon: '➶', group: 'toys', description: 'Launches at 45 degrees.', make: (x, y) => ({ entity: { type: 'cannon', x, y, angle: -45, power: 11 } }) },
  { id: 'balloon', name: 'Balloon', icon: '🎈', group: 'toys', description: 'Lifts the rider for three seconds.', make: (x, y) => ({ entity: { type: 'balloon', x, y, lift: 0.3, duration: 120 } }) },
  { id: 'platformV', name: 'Lift', icon: '⇕', group: 'toys', description: 'Platform moving up and down.', make: (x, y) => ({ entity: { type: 'platform', x, y, w: 70, dx: 0, dy: 60, period: 200 } }) },
  { id: 'platformH', name: 'Shuttle', icon: '⇔', group: 'toys', description: 'Platform moving side to side.', make: (x, y) => ({ entity: { type: 'platform', x, y, w: 70, dx: 90, dy: 0, period: 220 } }) },
  { id: 'wall', name: 'Breakable wall', icon: '▮', group: 'hazards', description: 'Shatters when hit fast.', make: (x, y) => ({ entity: { type: 'wall', x1: x, y1: y - 40, x2: x, y2: y + 40, threshold: 4 } }) },
  { id: 'gear', name: 'Gear', icon: '⚙', group: 'hazards', description: 'Spinning toothed wheel.', make: (x, y) => ({ entity: { type: 'gear', x, y, radius: 60, teeth: 8, speed: 0.02 } }) },
  { id: 'train', name: 'Train', icon: '🚃', group: 'hazards', description: 'Shuttles back and forth.', make: (x, y) => ({ entity: { type: 'train', x, y, w: 110, h: 28, speed: 2, range: 300 } }) },
  { id: 'domino', name: 'Domino', icon: '▯', group: 'props', description: 'Knock it over.', make: (x, y) => ({ prop: { kind: 'domino', x, y: y - 12, width: 4, height: 24, mass: 0.5 } }) },
  { id: 'crate', name: 'Crate', icon: '▣', group: 'props', description: 'A pushable box.', make: (x, y) => ({ prop: { kind: 'crate', x, y: y - 10, width: 20, height: 20, mass: 1.5 } }) },
  { id: 'tnt', name: 'TNT', icon: '✸', group: 'props', description: 'Explodes on a hard hit.', make: (x, y) => ({ prop: { kind: 'tnt', x, y: y - 9, width: 18, height: 18, mass: 1, explosive: true } }) },
  { id: 'boulder', name: 'Boulder', icon: '●', group: 'props', description: 'Heavy and rolls.', make: (x, y) => ({ prop: { kind: 'boulder', x, y: y - 14, radius: 14, mass: 6 } }) },
  { id: 'cart', name: 'Cart', icon: '▭', group: 'props', description: 'Low, slippery, fast.', make: (x, y) => ({ prop: { kind: 'cart', x, y: y - 6, width: 34, height: 12, mass: 1.2, friction: 0.05 } }) },
];

export function objectKind(id: ObjectKind): ObjectKindDef {
  return OBJECT_KINDS.find((k) => k.id === id)!;
}

/** Anchor point used for hover/erase on a placed entity. */
export function entityAnchor(def: EntityDef): { x: number; y: number } {
  switch (def.type) {
    case 'wall':
      return { x: (def.x1 + def.x2) / 2, y: (def.y1 + def.y2) / 2 };
    case 'collapse':
      return { x: (def.x1 + def.x2) / 2, y: (def.y1 + def.y2) / 2 };
    case 'fan':
      return { x: def.x + def.w / 2, y: def.y + def.h / 2 };
    case 'avalanche':
      return { x: def.startX, y: 0 };
    case 'pendulum':
      return { x: def.x, y: def.y + def.length };
    default:
      return { x: def.x, y: def.y };
  }
}
